File: core/automation/src/main/kotlin/cool/jacoblin/particeps/core/automation/Timers.kt
package cool.jacoblin.particeps.core.automation

import cool.jacoblin.particeps.core.definition.AutomationSchedule
import cool.jacoblin.particeps.core.definition.DurationClock
import cool.jacoblin.particeps.core.definition.OccurrenceAutomation
import java.security.SecureRandom
import java.time.Instant
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.ZoneId
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter

sealed interface TimerTarget {
    data class CalendarUtc(val utcMillis: Long) : TimerTarget {
        init { require(utcMillis >= 0) { "Calendar target must be non-negative" } }
    }
    data class ActiveElapsed(val elapsedNanos: Long) : TimerTarget {
        init { require(elapsedNanos >= 0) { "Active target must be non-negative" } }
    }
    data class SameBootMonotonic(val bootSessionId: String, val elapsedRealtimeNanos: Long) : TimerTarget {
        init {
            require(bootSessionId.isNotBlank()) { "Boot session ID must not be blank" }
            require(elapsedRealtimeNanos >= 0) { "Monotonic target must be non-negative" }
        }
    }
}
data class DurableTimer(
    val id: String,
    val automationId: String,
    val generation: ULong,
    val causalSequence: Long,
    val producerKey: String,
    val target: TimerTarget,
    val logicalDeadlineUtcMillis: Long?,
    val expiresAtUtcMillis: Long?,
) {
    init {
        require(TIMER_ID.matches(id)) { "Invalid timer ID" }
        require(AUTOMATION_ID.matches(automationId)) { "Invalid timer automation ID" }
        require(generation > 0uL) { "Timer generation must be positive" }
        require(causalSequence > 0) { "Timer causal sequence must be positive" }
        require(producerKey.length in 1..160 && '\u0000' !in producerKey) { "Invalid timer producer key" }
        require(logicalDeadlineUtcMillis == null || logicalDeadlineUtcMillis >= 0) { "Invalid logical deadline" }
        require(expiresAtUtcMillis == null || expiresAtUtcMillis >= 0) { "Invalid timer expiry" }
        require(
            logicalDeadlineUtcMillis == null || expiresAtUtcMillis == null ||
                expiresAtUtcMillis >= logicalDeadlineUtcMillis,
        ) { "Timer expiry precedes its deadline" }
    }

    private companion object {
        val TIMER_ID = Regex("[0-9a-f]{64}")
        val AUTOMATION_ID = Regex("[a-z0-9][a-z0-9-]{2,63}")
    }
}
sealed interface TimerIntent {
    data class Schedule(val timer: DurableTimer) : TimerIntent
    data class Retire(val timerId: String, val generation: ULong) : TimerIntent
}
data class MaterializedTimerSummary(val producerKey: String, val selectedUtcMillis: Long, val terminal: Boolean)
data class TimerProductionRequest(
    val configurationSha256: String,
    val automation: OccurrenceAutomation,
    val schedule: AutomationSchedule,
    val clock: ReducerClock,
    val studyStartUtcMillis: Long,
    val studyDeadlineUtcMillis: Long,
    val causalSequence: Long,
    val currentGeneration: ULong,
    val sessionState: StudySessionState,
    val pendingTimer: DurableTimer?,
    val materialized: List<MaterializedTimerSummary>,
) {
    init {
        require(studyStartUtcMillis >= 0 && studyDeadlineUtcMillis > studyStartUtcMillis) { "Invalid study interval" }
        require(causalSequence > 0) { "Timer production requires a causal sequence" }
        require(currentGeneration < ULong.MAX_VALUE) { "Timer generation overflow" }
        require(materialized.map { it.producerKey }.distinct().size == materialized.size) {
            "Duplicate materialized producer key"
        }
        require(pendingTimer == null || pendingTimer.automationId == automation.id) { "Pending timer automation mismatch" }
    }
}
sealed interface TimerProductionResult {
    data class Materialized(val timer: DurableTimer) : TimerProductionResult
    data object Deferred : TimerProductionResult
    data object Exhausted : TimerProductionResult
}
fun interface TimerProducer {
    fun produce(request: TimerProductionRequest): TimerProductionResult
}
fun interface BoundedRandomSource { fun nextInt(boundExclusive: Int): Int }
class SecureRandomSource(private val secureRandom: SecureRandom = SecureRandom()) : BoundedRandomSource {
    override fun nextInt(boundExclusive: Int): Int = secureRandom.nextInt(boundExclusive)
}
class StandardTimerProducer : TimerProducer {
    override fun produce(request: TimerProductionRequest): TimerProductionResult {
        request.pendingTimer?.let { return TimerProductionResult.Materialized(it) }
        if (request.sessionState != StudySessionState.RUNNING) return TimerProductionResult.Deferred
        val schedule = request.schedule
        require(schedule !is AutomationSchedule.RandomWindow) { "Random schedules require RandomWindowTimerProducer" }
        val produced = when (schedule) {
            is AutomationSchedule.OneTime -> produceOneTime(request, schedule)
            is AutomationSchedule.Interval -> produceInterval(request, schedule)
            is AutomationSchedule.DailyLocal -> produceDaily(request, schedule)
            is AutomationSchedule.RandomWindow -> error("Checked above")
        }
        return produced?.let(TimerProductionResult::Materialized) ?: TimerProductionResult.Exhausted
    }

    private fun produceOneTime(request: TimerProductionRequest, schedule: AutomationSchedule.OneTime): DurableTimer? {
        val producerKey = "one-time"
        if (request.materialized.any { it.producerKey == producerKey }) return null
        return when (schedule.clock) {
            DurationClock.CALENDAR_TIME -> {
                val deadline = checkedAddMillis(request.studyStartUtcMillis, schedule.offsetMinutes.toLong() * MILLIS_PER_MINUTE)
                if (!calendarDeadlineEligible(request, deadline)) null else calendarTimer(request, producerKey, deadline)
            }
            DurationClock.ACTIVE_RUNNING_TIME -> {
                val target = checkedMultiplyNanos(schedule.offsetMinutes.toLong(), SECONDS_PER_MINUTE)
                activeTimer(request, producerKey, target)
            }
        }
    }

    private fun produceInterval(request: TimerProductionRequest, schedule: AutomationSchedule.Interval): DurableTimer? {
        var ordinal = 0L
        val materializedKeys = request.materialized.mapTo(hashSetOf(), MaterializedTimerSummary::producerKey)
        while (ordinal < request.automation.maximumActivations.toLong()) {
            val producerKey = "interval:$ordinal"
            if (producerKey !in materializedKeys) {
                val offsetMinutes = addExact(
                    schedule.startOffsetMinutes.toLong(),
                    multiplyExact(ordinal, schedule.intervalMinutes.toLong()),
                )
                return when (schedule.clock) {
                    DurationClock.CALENDAR_TIME -> {
                        val deadline = checkedAddMillis(request.studyStartUtcMillis, multiplyExact(offsetMinutes, MILLIS_PER_MINUTE))
                        if (deadline >= request.studyDeadlineUtcMillis) null
                        else if (minimumExpiry(request, deadline) <= request.clock.now.wallTimeUtcMillis) {
                            ordinal++
                            continue
                        } else {
                            calendarTimer(request, producerKey, deadline)
                        }
                    }
                    DurationClock.ACTIVE_RUNNING_TIME -> {
                        val target = checkedMultiplyNanos(offsetMinutes, SECONDS_PER_MINUTE)
                        val studyDurationNanos = multiplyExact(
                            request.studyDeadlineUtcMillis - request.studyStartUtcMillis,
                            NANOS_PER_MILLI,
                        )
                        if (target >= studyDurationNanos) {
                            null
                        } else {
                            activeTimer(request, producerKey, target)
                        }
                    }
                }
            }
            ordinal++
        }
        return null
    }

    private fun produceDaily(request: TimerProductionRequest, schedule: AutomationSchedule.DailyLocal): DurableTimer? {
        val zone = ZoneId.of(request.clock.zoneId)
        val localTime = LocalTime.parse(schedule.localTime, DateTimeFormatter.ofPattern("HH:mm"))
        val nowInstant = Instant.ofEpochMilli(request.clock.now.wallTimeUtcMillis)
        val lastInstant = Instant.ofEpochMilli(request.studyDeadlineUtcMillis - 1)
        var date = maxOf(
            Instant.ofEpochMilli(request.studyStartUtcMillis).atZone(zone).toLocalDate(),
            nowInstant.atZone(zone).toLocalDate(),
        )
        val endDate = lastInstant.atZone(zone).toLocalDate()
        val materializedKeys = request.materialized.mapTo(hashSetOf(), MaterializedTimerSummary::producerKey)
        while (!date.isAfter(endDate)) {
            val producerKey = "daily:$date"
            if (producerKey !in materializedKeys) {
                val instant = firstInstant(LocalDateTime.of(date, localTime), zone)
                if (instant != null) {
                    val deadline = instant.toEpochMilli()
                    val expiry = minimumExpiry(request, deadline)
                    if (deadline >= request.studyStartUtcMillis && deadline < request.studyDeadlineUtcMillis &&
                        expiry > request.clock.now.wallTimeUtcMillis
                    ) {
                        return calendarTimer(request, producerKey, deadline)
                    }
                }
            }
            date = date.plusDays(1)
        }
        return null
    }

    private fun activeTimer(request: TimerProductionRequest, producerKey: String, targetNanos: Long): DurableTimer? {
        val studyDurationNanos = multiplyExact(
            request.studyDeadlineUtcMillis - request.studyStartUtcMillis,
            NANOS_PER_MILLI,
        )
        if (targetNanos < 0 || targetNanos >= studyDurationNanos) return null
        return DurableTimer(
            id = DeterministicIds.timerId(request.configurationSha256, request.automation.id, producerKey),
            automationId = request.automation.id,
            generation = request.currentGeneration + 1uL,
            causalSequence = request.causalSequence,
            producerKey = producerKey,
            target = TimerTarget.ActiveElapsed(targetNanos),
            logicalDeadlineUtcMillis = null,
            expiresAtUtcMillis = null,
        )
    }

    private fun calendarTimer(request: TimerProductionRequest, producerKey: String, deadline: Long): DurableTimer = DurableTimer(
        id = DeterministicIds.timerId(request.configurationSha256, request.automation.id, producerKey),
        automationId = request.automation.id,
        generation = request.currentGeneration + 1uL,
        causalSequence = request.causalSequence,
        producerKey = producerKey,
        target = TimerTarget.CalendarUtc(deadline),
        logicalDeadlineUtcMillis = deadline,
        expiresAtUtcMillis = minimumExpiry(request, deadline),
    )

    private fun calendarDeadlineEligible(request: TimerProductionRequest, deadline: Long): Boolean =
        deadline in request.studyStartUtcMillis until request.studyDeadlineUtcMillis &&
            minimumExpiry(request, deadline) > request.clock.now.wallTimeUtcMillis
}
class RandomWindowTimerProducer(private val random: BoundedRandomSource) : TimerProducer {
    override fun produce(request: TimerProductionRequest): TimerProductionResult {
        request.pendingTimer?.let { return TimerProductionResult.Materialized(it) }
        if (request.sessionState != StudySessionState.RUNNING) return TimerProductionResult.Deferred
        val schedule = request.schedule as? AutomationSchedule.RandomWindow
            ?: throw IllegalArgumentException("RandomWindowTimerProducer requires a random_window schedule")
        val candidates = RandomWindowEligibility.next(request, schedule) ?: return TimerProductionResult.Exhausted
        val selectedIndex = random.nextInt(candidates.utcMillis.size)
        require(selectedIndex in candidates.utcMillis.indices) { "Random source returned an out-of-range index" }
        return TimerProductionResult.Materialized(candidates.timer(request, candidates.utcMillis[selectedIndex]))
    }
}

/** Closed-world validation used by both the coordinator producer and reducer replay. */
object ScheduleTimerValidator {
    fun requireValid(request: TimerProductionRequest, timer: DurableTimer) {
        require(request.pendingTimer == null) { "A schedule timer is already pending" }
        when (val schedule = request.schedule) {
            is AutomationSchedule.RandomWindow -> {
                val candidates = RandomWindowEligibility.next(request, schedule)
                    ?: throw IllegalArgumentException("Random schedule is exhausted")
                require(timer.producerKey == candidates.producerKey) { "Random timer producer key is not next" }
                val selected = (timer.target as? TimerTarget.CalendarUtc)?.utcMillis
                    ?: throw IllegalArgumentException("Random timer must use a calendar target")
                require(selected in candidates.utcMillis) { "Random timer selection is outside the eligible set" }
                require(timer == candidates.timer(request, selected)) { "Random timer fields do not match its selection" }
            }
            else -> {
                val produced = StandardTimerProducer().produce(request)
                val expected = (produced as? TimerProductionResult.Materialized)?.timer
                    ?: throw IllegalArgumentException("Standard schedule is not eligible")
                require(timer == expected) { "Standard timer does not match the signed schedule" }
            }
        }
    }
}

private data class RandomWindowCandidates(val producerKey: String, val utcMillis: List<Long>) {
    init { require(utcMillis.isNotEmpty()) { "Random candidate set must not be empty" } }

    fun timer(request: TimerProductionRequest, selected: Long): DurableTimer = DurableTimer(
        id = DeterministicIds.timerId(request.configurationSha256, request.automation.id, producerKey),
        automationId = request.automation.id,
        generation = request.currentGeneration + 1uL,
        causalSequence = request.causalSequence,
        producerKey = producerKey,
        target = TimerTarget.CalendarUtc(selected),
        logicalDeadlineUtcMillis = selected,
        expiresAtUtcMillis = minimumExpiry(request, selected),
    )
}

private object RandomWindowEligibility {
    fun next(request: TimerProductionRequest, schedule: AutomationSchedule.RandomWindow): RandomWindowCandidates? {
        if (request.materialized.size >= schedule.maximumOccurrencesTotal) return null
        val zone = ZoneId.of(request.clock.zoneId)
        val nowMillis = request.clock.now.wallTimeUtcMillis
        val earliestDate = Instant.ofEpochMilli(request.studyStartUtcMillis).atZone(zone).toLocalDate()
        val latestDate = Instant.ofEpochMilli(request.studyDeadlineUtcMillis - 1).atZone(zone).toLocalDate()
        val materializedKeys = request.materialized.mapTo(hashSetOf(), MaterializedTimerSummary::producerKey)
        val chronologicalFloor = request.materialized.maxOfOrNull(MaterializedTimerSummary::selectedUtcMillis)
        val separationMillis = schedule.minimumSeparationMinutes.toLong() * MILLIS_PER_MINUTE
        var date = earliestDate
        while (!date.isAfter(latestDate)) {
            val datePrefix = "random:$date:"
            val remainingDaily = schedule.maximumOccurrencesPerDay -
                request.materialized.count { it.producerKey.startsWith(datePrefix) }
            if (remainingDaily > 0) {
                schedule.localWindows.forEachIndexed { windowIndex, window ->
                    val startMinute = parseMinute(window.startLocalTime)
                    val endMinute = parseMinute(window.endLocalTime)
                    for (ordinal in 0 until schedule.occurrencesPerWindow) {
                        val totalRemaining = schedule.maximumOccurrencesTotal - request.materialized.size
                        if (totalRemaining <= 0) return null
                        val producerKey = "random:$date:$windowIndex:$ordinal"
                        if (producerKey in materializedKeys) continue
                        val laterUnmaterialized = ((ordinal + 1) until schedule.occurrencesPerWindow).count { later ->
                            "random:$date:$windowIndex:$later" !in materializedKeys
                        }
                        val reservedLater = minOf(laterUnmaterialized, remainingDaily - 1, totalRemaining - 1)
                        val latestMinute = endMinute - 1 - reservedLater * schedule.minimumSeparationMinutes
                        if (latestMinute < startMinute) continue
                        val preceding = if (ordinal == 0) null else request.materialized.singleOrNull {
                            it.producerKey == "random:$date:$windowIndex:${ordinal - 1}"
                        }?.selectedUtcMillis
                        val eligible = ArrayList<Long>(latestMinute - startMinute + 1)
                        for (minute in startMinute..latestMinute) {
                            val local = LocalDateTime.of(date, LocalTime.of(minute / 60, minute % 60))
                            val instant = firstInstant(local, zone) ?: continue
                            val candidate = instant.toEpochMilli()
                            if (candidate < request.studyStartUtcMillis || candidate < nowMillis ||
                                candidate >= request.studyDeadlineUtcMillis ||
                                (chronologicalFloor != null && candidate <= chronologicalFloor) ||
                                (preceding != null && candidate - preceding < separationMillis) ||
                                request.materialized.any {
                                    absoluteDifference(candidate, it.selectedUtcMillis) < separationMillis
                                }
                            ) continue
                            eligible += candidate
                        }
                        if (eligible.isNotEmpty()) return RandomWindowCandidates(producerKey, eligible)
                    }
                }
            }
            date = date.plusDays(1)
        }
        return null
    }
}

private fun minimumExpiry(request: TimerProductionRequest, deadlineUtcMillis: Long): Long = minOf(
    checkedAddMillis(deadlineUtcMillis, request.automation.availabilitySeconds.toLong() * MILLIS_PER_SECOND),
    request.studyDeadlineUtcMillis,
)

private fun firstInstant(localDateTime: LocalDateTime, zone: ZoneId): Instant? {
    val offsets = zone.rules.getValidOffsets(localDateTime)
    if (offsets.isEmpty()) return null
    return offsets.map { offset: ZoneOffset -> localDateTime.toInstant(offset) }.minOrNull()
}

private fun parseMinute(value: String): Int = value.substring(0, 2).toInt() * 60 + value.substring(3).toInt()

private fun absoluteDifference(left: Long, right: Long): Long = when {
    left >= right -> left - right
    right >= left -> right - left
    else -> Long.MAX_VALUE
}

private fun checkedAddMillis(left: Long, right: Long): Long = addExact(left, right)
private fun checkedMultiplyNanos(value: Long, secondsPerUnit: Long): Long =
    multiplyExact(multiplyExact(value, secondsPerUnit), NANOS_PER_SECOND)

private fun addExact(left: Long, right: Long): Long = Math.addExact(left, right)
private fun multiplyExact(left: Long, right: Long): Long = Math.multiplyExact(left, right)

private const val MILLIS_PER_SECOND = 1_000L
private const val MILLIS_PER_MINUTE = 60_000L
private const val NANOS_PER_MILLI = 1_000_000L
private const val NANOS_PER_SECOND = 1_000_000_000L
private const val SECONDS_PER_MINUTE = 60L

File: web/src/lib/particeps/automation/timers.ts
import type { AutomationSchedule } from '../types';
import { timerId } from './checkpoint';
import type { DurableTimer, TimerProductionRequest } from './types';

const MINUTE_MS = 60_000;
const SECOND_MS = 1_000;
const NANO = 1_000_000_000n;

/**
 * Validates committed timer materialization without sampling entropy. Random
 * selections must be the next signed producer slot and a member of its exact
 * eligible UTC-minute set.
 */
export function requireValidScheduleTimer(request: TimerProductionRequest, timer: DurableTimer): void {
  if (request.pending_timer !== null) throw new Error('automation_schedule_timer_pending');
  let expected: DurableTimer | null;
  if (request.schedule.type === 'random_window') {
    const candidates = nextRandomCandidates(request, request.schedule);
    if (!candidates) throw new Error('automation_random_schedule_exhausted');
    if (timer.producer_key !== candidates.producerKey) throw new Error('automation_random_producer_not_next');
    if (timer.target.type !== 'CALENDAR_UTC' || !candidates.utcMillis.includes(timer.target.utc_millis)) {
      throw new Error('automation_random_selection_ineligible');
    }
    expected = calendarTimer(request, candidates.producerKey, timer.target.utc_millis);
  } else {
    expected = produceStandard(request);
    if (!expected) throw new Error('automation_standard_schedule_ineligible');
  }
  if (!timerEquals(timer, expected)) throw new Error('automation_timer_signed_schedule_mismatch');
}

/** Deterministic preview/runtime adapter for every non-random signed schedule. */
export function produceStandardScheduleTimer(request: TimerProductionRequest): DurableTimer | null {
  if (request.schedule.type === 'random_window' || request.pending_timer !== null) return null;
  return produceStandard(request);
}

function produceStandard(request: TimerProductionRequest): DurableTimer | null {
  if (request.pending_timer) return request.pending_timer;
  if (request.session_state !== 'RUNNING') return null;
  const schedule = request.schedule;
  if (schedule.type === 'one_time') {
    const producer = 'one-time';
    if (hasKey(request, producer)) return null;
    if (schedule.clock === 'CALENDAR_TIME') {
      const deadline = request.study_start_utc_millis + schedule.offset_minutes * MINUTE_MS;
      return calendarEligible(request, deadline) ? calendarTimer(request, producer, deadline) : null;
    }
    return activeTimer(request, producer, BigInt(schedule.offset_minutes) * 60n * NANO);
  }
  if (schedule.type === 'interval') {
    const keys = new Set(request.materialized.map((item) => item.producer_key));
    for (let ordinal = 0; ordinal < request.automation.maximum_activations; ordinal += 1) {
      const producer = `interval:${ordinal}`;
      if (keys.has(producer)) continue;
      const offset = schedule.start_offset_minutes + ordinal * schedule.interval_minutes;
      if (schedule.clock === 'CALENDAR_TIME') {
        const deadline = request.study_start_utc_millis + offset * MINUTE_MS;
        if (deadline >= request.study_deadline_utc_millis) return null;
        if (minimumExpiry(request, deadline) <= request.clock.now.wall_time_utc_millis) continue;
        return calendarTimer(request, producer, deadline);
      }
      return activeTimer(request, producer, BigInt(offset) * 60n * NANO);
    }
    return null;
  }
  if (schedule.type === 'daily_local') {
    let date = maxDate(localDate(request.study_start_utc_millis, request.clock.zone_id),
      localDate(request.clock.now.wall_time_utc_millis, request.clock.zone_id));
    const lastDate = localDate(request.study_deadline_utc_millis - 1, request.clock.zone_id);
    const keys = new Set(request.materialized.map((item) => item.producer_key));
    while (date <= lastDate) {
      const producer = `daily:${date}`;
      if (!keys.has(producer)) {
        const instant = firstInstant(date, schedule.local_time, request.clock.zone_id);
        if (instant !== null && calendarEligible(request, instant)) return calendarTimer(request, producer, instant);
      }
      date = nextDate(date);
    }
    return null;
  }
  throw new Error('automation_standard_producer_random_schedule');
}

function nextRandomCandidates(
  request: TimerProductionRequest,
  schedule: Extract<AutomationSchedule, { type: 'random_window' }>
): { producerKey: string; utcMillis: number[] } | null {
  if (request.session_state !== 'RUNNING' || request.materialized.length >= schedule.maximum_occurrences_total) return null;
  let date = localDate(request.study_start_utc_millis, request.clock.zone_id);
  const lastDate = localDate(request.study_deadline_utc_millis - 1, request.clock.zone_id);
  const keys = new Set(request.materialized.map((item) => item.producer_key));
  const chronologicalFloor = request.materialized.length
    ? Math.max(...request.materialized.map((item) => item.selected_utc_millis)) : null;
  const separation = schedule.minimum_separation_minutes * MINUTE_MS;
  while (date <= lastDate) {
    const prefix = `random:${date}:`;
    const remainingDaily = schedule.maximum_occurrences_per_day -
      request.materialized.filter((item) => item.producer_key.startsWith(prefix)).length;
    if (remainingDaily > 0) {
      for (const [windowIndex, window] of schedule.local_windows.entries()) {
        const start = parseMinute(window.start_local_time);
        const end = parseMinute(window.end_local_time);
        for (let ordinal = 0; ordinal < schedule.occurrences_per_window; ordinal += 1) {
          const totalRemaining = schedule.maximum_occurrences_total - request.materialized.length;
          if (totalRemaining <= 0) return null;
          const producerKey = `random:${date}:${windowIndex}:${ordinal}`;
          if (keys.has(producerKey)) continue;
          let later = 0;
          for (let candidate = ordinal + 1; candidate < schedule.occurrences_per_window; candidate += 1) {
            if (!keys.has(`random:${date}:${windowIndex}:${candidate}`)) later += 1;
          }
          const reserved = Math.min(later, remainingDaily - 1, totalRemaining - 1);
          const latest = end - 1 - reserved * schedule.minimum_separation_minutes;
          if (latest < start) continue;
          const preceding = ordinal === 0 ? null : request.materialized.find((item) =>
            item.producer_key === `random:${date}:${windowIndex}:${ordinal - 1}`)?.selected_utc_millis ?? null;
          const eligible: number[] = [];
          for (let minute = start; minute <= latest; minute += 1) {
            const candidate = firstInstant(date, minuteText(minute), request.clock.zone_id);
            if (candidate === null || candidate < request.study_start_utc_millis ||
              candidate < request.clock.now.wall_time_utc_millis || candidate >= request.study_deadline_utc_millis ||
              (chronologicalFloor !== null && candidate <= chronologicalFloor) ||
              (preceding !== null && candidate - preceding < separation) ||
              request.materialized.some((item) => Math.abs(candidate - item.selected_utc_millis) < separation)) continue;
            eligible.push(candidate);
          }
          if (eligible.length) return { producerKey, utcMillis: eligible };
        }
      }
    }
    date = nextDate(date);
  }
  return null;
}

function activeTimer(request: TimerProductionRequest, producer: string, target: bigint): DurableTimer | null {
  const duration = BigInt(request.study_deadline_utc_millis - request.study_start_utc_millis) * 1_000_000n;
  if (target < 0n || target >= duration) return null;
  return {
    id: timerId(request.configuration_sha256, request.automation.id, producer), automation_id: request.automation.id,
    generation: request.current_generation + 1n, causal_sequence: request.causal_sequence, producer_key: producer,
    target: { type: 'ACTIVE_ELAPSED', elapsed_nanos: target }, logical_deadline_utc_millis: null,
    expires_at_utc_millis: null
  };
}

function calendarTimer(request: TimerProductionRequest, producer: string, deadline: number): DurableTimer {
  return {
    id: timerId(request.configuration_sha256, request.automation.id, producer), automation_id: request.automation.id,
    generation: request.current_generation + 1n, causal_sequence: request.causal_sequence, producer_key: producer,
    target: { type: 'CALENDAR_UTC', utc_millis: deadline }, logical_deadline_utc_millis: deadline,
    expires_at_utc_millis: minimumExpiry(request, deadline)
  };
}

function minimumExpiry(request: TimerProductionRequest, deadline: number): number {
  return Math.min(deadline + request.automation.availability_seconds * SECOND_MS, request.study_deadline_utc_millis);
}

function calendarEligible(request: TimerProductionRequest, deadline: number): boolean {
  return deadline >= request.study_start_utc_millis && deadline < request.study_deadline_utc_millis &&
    minimumExpiry(request, deadline) > request.clock.now.wall_time_utc_millis;
}

function hasKey(request: TimerProductionRequest, producer: string): boolean {
  return request.materialized.some((item) => item.producer_key === producer);
}

function firstInstant(date: string, localTime: string, zone: string): number | null {
  const [year, month, day] = date.split('-').map(Number);
  const [hour, minute] = localTime.split(':').map(Number);
  const localAsUtc = Date.UTC(year, month - 1, day, hour, minute, 0, 0);
  const offsets = new Set([-172_800_000, -86_400_000, 0, 86_400_000, 172_800_000]
    .map((delta) => zoneOffsetMillis(localAsUtc + delta, zone)));
  const candidates = [...offsets].map((offset) => localAsUtc - offset).filter((candidate) => {
    const parts = zonedParts(candidate, zone);
    return parts.year === year && parts.month === month && parts.day === day && parts.hour === hour && parts.minute === minute;
  });
  return candidates.length ? Math.min(...candidates) : null;
}

function zoneOffsetMillis(instant: number, zone: string): number {
  const second = Math.trunc(instant / 1_000) * 1_000;
  const parts = zonedParts(second, zone);
  return Date.UTC(parts.year, parts.month - 1, parts.day, parts.hour, parts.minute, parts.second) - second;
}

function zonedParts(instant: number, zone: string): Record<'year' | 'month' | 'day' | 'hour' | 'minute' | 'second', number> {
  const formatter = new Intl.DateTimeFormat('en-CA', {
    timeZone: zone, calendar: 'iso8601', numberingSystem: 'latn', hourCycle: 'h23',
    year: 'numeric', month: '2-digit', day: '2-digit', hour: '2-digit', minute: '2-digit', second: '2-digit'
  });
  const result = {} as Record<'year' | 'month' | 'day' | 'hour' | 'minute' | 'second', number>;
  for (const part of formatter.formatToParts(new Date(instant))) {
    if (part.type in { year: 1, month: 1, day: 1, hour: 1, minute: 1, second: 1 }) {
      result[part.type as keyof typeof result] = Number(part.value);
    }
  }
  if (Object.keys(result).length !== 6) throw new Error('automation_zone_format_failure');
  return result;
}

function localDate(instant: number, zone: string): string {
  const parts = zonedParts(instant, zone);
  return `${parts.year.toString().padStart(4, '0')}-${parts.month.toString().padStart(2, '0')}-${parts.day.toString().padStart(2, '0')}`;
}

function nextDate(value: string): string {
  const [year, month, day] = value.split('-').map(Number);
  const date = new Date(Date.UTC(year, month - 1, day + 1));
  return `${date.getUTCFullYear().toString().padStart(4, '0')}-${(date.getUTCMonth() + 1).toString().padStart(2, '0')}-${date.getUTCDate().toString().padStart(2, '0')}`;
}

const maxDate = (left: string, right: string): string => left > right ? left : right;
const parseMinute = (value: string): number => Number(value.slice(0, 2)) * 60 + Number(value.slice(3));
const minuteText = (value: number): string => `${Math.trunc(value / 60).toString().padStart(2, '0')}:${(value % 60).toString().padStart(2, '0')}`;

function timerEquals(left: DurableTimer, right: DurableTimer): boolean {
  return left.id === right.id && left.automation_id === right.automation_id && left.generation === right.generation &&
    left.causal_sequence === right.causal_sequence && left.producer_key === right.producer_key &&
    JSON.stringify(left.target, (_, value) => typeof value === 'bigint' ? value.toString() : value) ===
      JSON.stringify(right.target, (_, value) => typeof value === 'bigint' ? value.toString() : value) &&
    left.logical_deadline_utc_millis === right.logical_deadline_utc_millis &&
    left.expires_at_utc_millis === right.expires_at_utc_millis;
}
